crates/matrix-sdk/src/widget/machine/mod.rs

```diff
@@ -616,10 +651,94 @@ impl WidgetMachine {
+                        StateEventFilter::WithTypeAndStateKey(event_type, state_key) => {
+                            ReadStateRequest {
+                                event_type: event_type.clone(),
+                                state_key: StateKeySelector::Key(state_key.clone()),
+                            }
+                        }
+                    });
+                    request.then(move |result, machine| {
+                        if initial_state_pending.is_cancelled() {
+                            Vec::new()
+                        } else {
+                            let mut updates =
+                                mem::take(&mut machine.pending_state_updates).unwrap();
+                            updates.initial_state.push(result.unwrap_or_else(|e| {
+                                error!("Reading initial room state failed: {e}");
+                                Vec::new()
+                            }));
+                            // Once the number of initial state responses reaches `request_count`,
+                            // the initial state is complete and ready to be pushed
+                            if updates.initial_state.len() == request_count {
+                                iter::once({
+                                    let (_request, action) =
+                                        machine.send_to_widget_request(NotifyStateUpdate {
+                                            state: updates
+                                                .initial_state
+                                                .into_iter()
+                                                .flatten()
+                                                .collect(),
+                                        });
+                                    action
+                                })
+                                .chain(updates.postponed_updates.into_iter().map(|state| {
+                                    let (_request, action) =
+                                        machine.send_to_widget_request(NotifyStateUpdate { state });
+                                    action
+                                }))
+                                .flatten()
+                                .collect()
+                            } else {
+                                // Continue accumulating the initial state
+                                machine.pending_state_updates = Some(updates);
+                                Vec::new()
+                            }
+                        }
+                    });
+                    action
+                });
 
-                subscribe_required.then_some(Action::Subscribe).into_iter().chain(action).collect()
+                subscribe_required
+                    .then_some(Action::Subscribe)
+                    .into_iter()
+                    .chain(initial_state_actions)
+                    .chain(notify_action)
+                    .collect()
```

Comment:
I know this is pre-existing, but this chaining of iterators is unnecessary complicated: instead, let's use a mutable vector of `Action` and push into it as we need to (like I've done in commits in #4966).